### ISSUE TITLE
Stalker ARM incorrectly modifies register for mov pc, <reg>.

### ISSUE THREAD

**Issue by WorksButNotTested** [frida/frida-gum#455]
As can be seen from the following generated code, stalker ARM incorrectly assumes that all returns are either `pop` or `ldmia` instructions and advances the base register as if popping the return address.

```
0xbb9ca212 mov pc, r2
	0xbb6390ee	push {r0, r1, r2, r3, r4, r5, r6, r7, lr}
	0xbb6390f0	push.w {r8, sb, sl, fp, ip}
	0xbb6390f4	mrs r0, apsr
	0xbb6390f8	subs r1, r1, r1
	0xbb6390fa	add r2, sp, #0x38
	0xbb6390fc	push {r0, r1, r2}
	0xbb6390fe	mov sl, sp
	0xbb639100	and r0, sp, #7
	0xbb639104	sub sp, #8
	0xbb639106	add sp, r0
	0xbb639108	vpush {d16, d17, d18, d19, d20, d21, d22, d23, d24, d25, d26, d27, d28, d29, d30, d31}
	0xbb63910c	vpush {d0, d1, d2, d3, d4, d5, d6, d7}
	0xbb639110	ldr.w r1, [sl, #0x28]
	0xbb639114	adds r1, #0
	0xbb639116	orr r1, r1, #1
	0xbb63911a	ldr r0, [pc, #0]
	0xbb63911c	ldr.w lr, [pc, #0]
	0xbb639120	blx lr
	0xbb639122	ldr.w r1, [sl, #0x28]
	0xbb639126	adds r1, #0
	0xbb639128	orr r1, r1, #1
	0xbb63912c	ldr r0, [pc, #0]
	0xbb63912e	ldr.w lr, [pc, #0]
	0xbb639132	blx lr
	0xbb639134	vpop {d0, d1, d2, d3, d4, d5, d6, d7}
	0xbb639138	vpop {d16, d17, d18, d19, d20, d21, d22, d23, d24, d25, d26, d27, d28, d29, d30, d31}
	0xbb63913c	mov sp, sl
	0xbb63913e	pop {r0, r1, r2}
	0xbb639140	msr cpsr_fc, r0
	0xbb639144	pop.w {r8, sb, sl, fp, ip}
	0xbb639148	pop.w {r0, r1, r2, r3, r4, r5, r6, r7, lr}
==>	0xbb63914c	adds r2, #4
	0xbb63914e	push {r0, r1}
	0xbb639150	ldr r0, [pc, #0]
	0xbb639152	ldr r0, [r0]
	0xbb639154	str r0, [sp, #4]
	0xbb639156	pop {r0, pc}
```

**Comment by WorksButNotTested** [comment]
As can be seen from the following generated code, stalker ARM incorrectly assumes that all returns are either `pop` or `ldmia` instructions and advances the base register as if popping the return address.

```
0xbb9ca212 mov pc, r2
	0xbb6390ee	push {r0, r1, r2, r3, r4, r5, r6, r7, lr}
	0xbb6390f0	push.w {r8, sb, sl, fp, ip}
	0xbb6390f4	mrs r0, apsr
	0xbb6390f8	subs r1, r1, r1
	0xbb6390fa	add r2, sp, #0x38
	0xbb6390fc	push {r0, r1, r2}
	0xbb6390fe	mov sl, sp
	0xbb639100	and r0, sp, #7
	0xbb639104	sub sp, #8
	0xbb639106	add sp, r0
	0xbb639108	vpush {d16, d17, d18, d19, d20, d21, d22, d23, d24, d25, d26, d27, d28, d29, d30, d31}
	0xbb63910c	vpush {d0, d1, d2, d3, d4, d5, d6, d7}
	0xbb639110	ldr.w r1, [sl, #0x28]
	0xbb639114	adds r1, #0
	0xbb639116	orr r1, r1, #1
	0xbb63911a	ldr r0, [pc, #0]
	0xbb63911c	ldr.w lr, [pc, #0]
	0xbb639120	blx lr
	0xbb639122	ldr.w r1, [sl, #0x28]
	0xbb639126	adds r1, #0
	0xbb639128	orr r1, r1, #1
	0xbb63912c	ldr r0, [pc, #0]
	0xbb63912e	ldr.w lr, [pc, #0]
	0xbb639132	blx lr
	0xbb639134	vpop {d0, d1, d2, d3, d4, d5, d6, d7}
	0xbb639138	vpop {d16, d17, d18, d19, d20, d21, d22, d23, d24, d25, d26, d27, d28, d29, d30, d31}
	0xbb63913c	mov sp, sl
	0xbb63913e	pop {r0, r1, r2}
	0xbb639140	msr cpsr_fc, r0
	0xbb639144	pop.w {r8, sb, sl, fp, ip}
	0xbb639148	pop.w {r0, r1, r2, r3, r4, r5, r6, r7, lr}
==>	0xbb63914c	adds r2, #4
	0xbb63914e	push {r0, r1}
	0xbb639150	ldr r0, [pc, #0]
	0xbb639152	ldr r0, [r0]
	0xbb639154	str r0, [sp, #4]
	0xbb639156	pop {r0, pc}
```

**Comment by WorksButNotTested** [comment]
The implementation of virtualization for ARM return instruction includes a `pop` parameter which is used to determine whether such an adjustment is required after the prologue is closed. This is missing from the Thumb code.

```c
static void
gum_exec_block_virtualize_arm_ret_insn (GumExecBlock * block,
                                        const GumBranchTarget * target,
                                        arm_cc cc,
                                        gboolean pop,
                                        guint16 mask,
                                        GumGeneratorContext * gc)
{
  GumExecCtx * ec = block->ctx;

  gum_exec_block_write_arm_handle_not_taken (block, target, cc, gc);

  gum_exec_block_arm_open_prolog (block, gc);

  if ((ec->sink_mask & GUM_EXEC) != 0)
    gum_exec_block_write_arm_exec_event_code (block, gc);

  gum_exec_block_write_arm_pop_stack_frame (block, target, gc);

  if ((ec->sink_mask & GUM_RET) != 0)
    gum_exec_block_write_arm_ret_event_code (block, target, gc);

  gum_exec_block_write_arm_call_replace_block (block, target, gc);

  gum_exec_block_arm_close_prolog (block, gc);

  /*
   * If the instruction we are virtualizing is a POP (or indeed LDMIA)
   * instruction, then as well as determining the location at which control flow
   * should continue, we must ensure we load any other registers in the register
   * list of the instruction from the stack. Lastly, we must increment that
   * stack pointer to remove the value of PC which would have been restored,
   * since we will instead control PC to continue execution of instrumented
   * code.
   */
  if (pop)
  {
    const GumBranchIndirectRegOffset * tv = &target->value.indirect_reg_offset;

    g_assert (target->type == GUM_TARGET_INDIRECT_REG_OFFSET);

    if (mask != 0)
      gum_arm_writer_put_ldmia_reg_mask (gc->arm_writer, tv->reg, mask);

    gum_arm_writer_put_add_reg_reg_imm (gc->arm_writer, tv->reg, tv->reg, 4);
  }

  gum_exec_block_write_arm_exec_generated_code (gc->arm_writer, block->ctx);
}
```

**Comment by WorksButNotTested** [comment]
Fixed by #457
